File: go/fixture/json_ledger_admin_client.go
// Copyright (c) 2026 Peaceful Studio OÜ
// SPDX-License-Identifier: Apache-2.0

package fixture

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// TokenProvider abstracts access-token retrieval for the admin client.
// OAuth2TokenProvider implements this interface; tests can substitute a
// static or scripted provider.
type TokenProvider interface {
	Token(ctx context.Context) (string, error)
}

// JsonLedgerAdminClient is a minimal HTTP client for the Canton JSON Ledger
// Admin API.
type JsonLedgerAdminClient struct {
	baseURL    string
	tokens     TokenProvider
	httpClient *http.Client
}

// JsonLedgerAdminClientOption mutates a JsonLedgerAdminClient at construction.
type JsonLedgerAdminClientOption func(*JsonLedgerAdminClient)

// WithHTTPClient overrides the default HTTP client used by the admin client.
func WithHTTPClient(c *http.Client) JsonLedgerAdminClientOption {
	return func(j *JsonLedgerAdminClient) {
		if c != nil {
			j.httpClient = c
		}
	}
}

// NewJsonLedgerAdminClient constructs a JsonLedgerAdminClient.
//
// baseURL is the JSON Ledger API origin, e.g. "http://localhost:11975".
// tokens supplies bearer tokens for the Authorization header.
func NewJsonLedgerAdminClient(baseURL string, tokens TokenProvider, opts ...JsonLedgerAdminClientOption) (*JsonLedgerAdminClient, error) {
	if strings.TrimSpace(baseURL) == "" {
		return nil, errors.New("admin: baseURL is required")
	}
	if tokens == nil {
		return nil, errors.New("admin: TokenProvider is required")
	}
	client := &JsonLedgerAdminClient{
		baseURL:    normalizeBaseURL(baseURL),
		tokens:     tokens,
		httpClient: http.DefaultClient,
	}
	for _, opt := range opts {
		opt(client)
	}
	return client, nil
}

type participantIDResponse struct {
	ParticipantID string `json:"participantId"`
}

// GetParticipantId calls GET /v2/parties/participant-id and returns the
// participantId field from the JSON response body.
func (c *JsonLedgerAdminClient) GetParticipantId(ctx context.Context) (string, error) {
	var out participantIDResponse
	if err := c.doJSON(ctx, http.MethodGet, "/v2/parties/participant-id", nil, &out); err != nil {
		return "", err
	}
	if strings.TrimSpace(out.ParticipantID) == "" {
		return "", errors.New("admin: response missing participantId")
	}
	return out.ParticipantID, nil
}

// AppSynchronizerAlias is the stable alias of the app-provider synchronizer.
const AppSynchronizerAlias = "app-synchronizer"

// ConnectedSynchronizer is one entry from GET /v2/state/connected-synchronizers.
type ConnectedSynchronizer struct {
	Alias      string
	ID         string
	Permission string
}

type connectedSynchronizersResponse struct {
	ConnectedSynchronizers []struct {
		SynchronizerAlias string `json:"synchronizerAlias"`
		SynchronizerID    string `json:"synchronizerId"`
		Permission        string `json:"permission"`
	} `json:"connectedSynchronizers"`
}

// GetConnectedSynchronizers lists the synchronizers the participant is
// connected to for party, via GET /v2/state/connected-synchronizers.
func (c *JsonLedgerAdminClient) GetConnectedSynchronizers(ctx context.Context, party string) ([]ConnectedSynchronizer, error) {
	if strings.TrimSpace(party) == "" {
		return nil, errors.New("admin: party is required")
	}
	var out connectedSynchronizersResponse
	path := "/v2/state/connected-synchronizers?party=" + url.QueryEscape(party)
	if err := c.doJSON(ctx, http.MethodGet, path, nil, &out); err != nil {
		return nil, err
	}
	result := make([]ConnectedSynchronizer, 0, len(out.ConnectedSynchronizers))
	for _, s := range out.ConnectedSynchronizers {
		if strings.TrimSpace(s.SynchronizerID) == "" {
			return nil, fmt.Errorf("admin: connected-synchronizers entry missing synchronizerId (alias=%q)", s.SynchronizerAlias)
		}
		result = append(result, ConnectedSynchronizer{Alias: s.SynchronizerAlias, ID: s.SynchronizerID, Permission: s.Permission})
	}
	return result, nil
}

// GetAppSynchronizerId returns the id of the connected synchronizer whose
// alias is AppSynchronizerAlias. Errors if absent (multi-sync off).
func (c *JsonLedgerAdminClient) GetAppSynchronizerId(ctx context.Context, party string) (string, error) {
	syncs, err := c.GetConnectedSynchronizers(ctx, party)
	if err != nil {
		return "", err
	}
	aliases := make([]string, 0, len(syncs))
	for _, s := range syncs {
		if s.Alias == AppSynchronizerAlias {
			return s.ID, nil
		}
		aliases = append(aliases, s.Alias)
	}
	return "", fmt.Errorf("admin: no connected synchronizer with alias %q (is multi-sync enabled?); connected: %v", AppSynchronizerAlias, aliases)
}

func (c *JsonLedgerAdminClient) doJSON(ctx context.Context, method, path string, body []byte, out any) error {
	respBody, err := doRequest(ctx, c.httpClient, c.tokens, c.baseURL, httpRequest{
		method:      method,
		path:        path,
		body:        body,
		contentType: "application/json",
		errPrefix:   "admin",
	})
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	if err := json.Unmarshal(respBody, out); err != nil {
		return fmt.Errorf("admin: decode %s %s response: %w", method, path, err)
	}
	return nil
}

File: csharp/src/Peaceful.Canton.Localnet.Testing/JsonLedgerAdminClient.cs
// Copyright (c) 2026 Peaceful Studio OÜ
// SPDX-License-Identifier: Apache-2.0

using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Peaceful.Canton.Localnet.Testing;

/// <summary>
/// Thin wrapper over the Canton JSON Ledger API v2. v0 implements only
/// <c>GET /v2/parties/participant-id</c>; subsequent slices (#9 onwards) will
/// add DAR upload, party allocation, and user binding.
/// </summary>
public sealed class JsonLedgerAdminClient
{
    private readonly HttpClient _httpClient;
    private readonly OAuth2TokenProvider _tokenProvider;
    private readonly ILogger<JsonLedgerAdminClient> _logger;

    /// <summary>
    /// Creates a client bound to a JSON Ledger API <see cref="HttpClient"/>.
    /// </summary>
    /// <param name="httpClient">
    /// Client whose <see cref="HttpClient.BaseAddress"/> is the JSON Ledger API
    /// root (e.g. <c>http://localhost:11975/</c>). Required; an unset base
    /// address throws.
    /// </param>
    /// <param name="tokenProvider">Supplies the bearer token for each request.</param>
    /// <param name="logger">Optional logger; defaults to a no-op logger.</param>
    /// <exception cref="ArgumentException">
    /// <paramref name="httpClient"/> has no <see cref="HttpClient.BaseAddress"/>.
    /// </exception>
    public JsonLedgerAdminClient(
        HttpClient httpClient,
        OAuth2TokenProvider tokenProvider,
        ILogger<JsonLedgerAdminClient>? logger = null)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _tokenProvider = tokenProvider ?? throw new ArgumentNullException(nameof(tokenProvider));
        _logger = logger ?? NullLogger<JsonLedgerAdminClient>.Instance;

        if (_httpClient.BaseAddress is null)
        {
            throw new ArgumentException(
                "HttpClient must have a BaseAddress set to the JSON Ledger API root (e.g. http://localhost:11975/).",
                nameof(httpClient));
        }
    }

    /// <summary>
    /// Calls <c>GET /v2/parties/participant-id</c> and returns the participant's
    /// identifier (e.g. <c>a-validator-1::1220...</c>). Throws
    /// <see cref="JsonLedgerApiException"/> if the server returns a non-success
    /// status or an empty body.
    /// </summary>
    public async Task<string> GetParticipantIdAsync(CancellationToken cancellationToken = default)
    {
        var token = await _tokenProvider.GetAccessTokenAsync(cancellationToken).ConfigureAwait(false);

        using var request = new HttpRequestMessage(HttpMethod.Get, "v2/parties/participant-id");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        _logger.LogDebug("GET {Uri}", new Uri(_httpClient.BaseAddress!, request.RequestUri!));

        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            throw new JsonLedgerApiException(
                $"GET v2/parties/participant-id returned {(int)response.StatusCode} {response.ReasonPhrase}: {body}",
                response.StatusCode,
                body);
        }

        var payload = await response.Content
            .ReadFromJsonAsync<ParticipantIdResponse>(cancellationToken: cancellationToken)
            .ConfigureAwait(false)
            ?? throw new JsonLedgerApiException(
                "GET v2/parties/participant-id returned an empty body.",
                response.StatusCode,
                string.Empty);

        if (string.IsNullOrWhiteSpace(payload.ParticipantId))
        {
            throw new JsonLedgerApiException(
                "GET v2/parties/participant-id returned a response with no participantId field.",
                response.StatusCode,
                string.Empty);
        }

        return payload.ParticipantId;
    }

    /// <summary>The stable alias of the app-provider synchronizer.</summary>
    public const string AppSynchronizerAlias = "app-synchronizer";

    /// <summary>
    /// Lists the synchronizers the participant is connected to for <paramref name="party"/>,
    /// via <c>GET /v2/state/connected-synchronizers</c>.
    /// </summary>
    public async Task<IReadOnlyList<ConnectedSynchronizer>> GetConnectedSynchronizersAsync(
        string party,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(party))
        {
            throw new ArgumentException("party must be a non-empty party id.", nameof(party));
        }

        var token = await _tokenProvider.GetAccessTokenAsync(cancellationToken).ConfigureAwait(false);
        var requestUri = $"v2/state/connected-synchronizers?party={Uri.EscapeDataString(party)}";

        using var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        _logger.LogDebug("GET {Uri}", new Uri(_httpClient.BaseAddress!, request.RequestUri!));

        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            throw new JsonLedgerApiException(
                $"GET {requestUri} returned {(int)response.StatusCode} {response.ReasonPhrase}: {body}",
                response.StatusCode,
                body);
        }

        var payload = await response.Content
            .ReadFromJsonAsync<ConnectedSynchronizersResponse>(cancellationToken: cancellationToken)
            .ConfigureAwait(false)
            ?? throw new JsonLedgerApiException(
                $"GET {requestUri} returned an empty body.",
                response.StatusCode,
                string.Empty);

        if (payload.ConnectedSynchronizers is null)
        {
            throw new JsonLedgerApiException(
                $"GET {requestUri} returned a response with no connectedSynchronizers field.",
                response.StatusCode,
                string.Empty);
        }

        foreach (var s in payload.ConnectedSynchronizers)
        {
            if (string.IsNullOrWhiteSpace(s.SynchronizerId))
            {
                throw new JsonLedgerApiException(
                    $"GET {requestUri} returned a connected-synchronizers entry with an empty synchronizerId (alias='{s.SynchronizerAlias}').",
                    response.StatusCode,
                    string.Empty);
            }
        }

        return payload.ConnectedSynchronizers
            .Select(s => new ConnectedSynchronizer(s.SynchronizerAlias, s.SynchronizerId, s.Permission))
            .ToList();
    }

    /// <summary>
    /// Returns the id of the connected synchronizer whose alias is
    /// <see cref="AppSynchronizerAlias"/>. Throws if it is absent (multi-sync off).
    /// </summary>
    public async Task<string> GetAppSynchronizerIdAsync(
        string party,
        CancellationToken cancellationToken = default)
    {
        var synchronizers = await GetConnectedSynchronizersAsync(party, cancellationToken).ConfigureAwait(false);
        var app = synchronizers.FirstOrDefault(s => string.Equals(s.Alias, AppSynchronizerAlias, StringComparison.Ordinal));
        if (app is null)
        {
            var aliases = string.Join(", ", synchronizers.Select(s => s.Alias));
            throw new JsonLedgerApiException(
                $"No connected synchronizer with alias '{AppSynchronizerAlias}' (is the multi-sync profile enabled?). Connected: [{aliases}].",
                System.Net.HttpStatusCode.NotFound,
                string.Empty);
        }

        return app.Id;
    }

    private sealed record ParticipantIdResponse(
        [property: JsonPropertyName("participantId")] string ParticipantId);

    private sealed record ConnectedSynchronizersResponse(
        [property: JsonPropertyName("connectedSynchronizers")] IReadOnlyList<ConnectedSynchronizerDto> ConnectedSynchronizers);

    private sealed record ConnectedSynchronizerDto(
        [property: JsonPropertyName("synchronizerAlias")] string SynchronizerAlias,
        [property: JsonPropertyName("synchronizerId")] string SynchronizerId,
        [property: JsonPropertyName("permission")] string? Permission);
}

/// <summary>
/// Raised when the JSON Ledger API returns a non-success status, an empty
/// response body, or a response that fails the v0 contract (missing fields).
/// <para>
/// This is a base type — <see cref="UserRightsGrantedWithoutLeaseException"/>
/// derives from it. A <c>catch</c> on this type catches the derived ones; xUnit's
/// <c>Assert.Throws&lt;T&gt;</c> and <c>ThrowsAsync&lt;T&gt;</c> match the exact
/// type and do not, so assert with <c>ThrowsAny</c> where a derived exception is
/// possible.
/// </para>
/// </summary>
public class JsonLedgerApiException : Exception
{
    /// <summary>
    /// Creates the exception with the originating <paramref name="statusCode"/>
    /// and raw <paramref name="responseBody"/> for diagnostics.
    /// </summary>
    /// <param name="message">Human-readable description of the failure.</param>
    /// <param name="statusCode">HTTP status the participant returned.</param>
    /// <param name="responseBody">Raw response body, or empty when none was read.</param>
    public JsonLedgerApiException(string message, System.Net.HttpStatusCode statusCode, string responseBody)
        : base(message)
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
    }

    /// <summary>HTTP status the participant returned.</summary>
    public System.Net.HttpStatusCode StatusCode { get; }

    /// <summary>Raw response body, or empty when none was read.</summary>
    public string ResponseBody { get; }
}
